[FLINK-20153] Add documentation for BATCH execution mode

aljoscha opened the pull request:
This adds documentation for the new `BATCH` execution mode. We also explain `STREAMING` execution mode because there is no central page that explains the basic behavior, so far.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

Documentation only.

alpinegizmo commented on docs/dev/datastream_execution_mode.md:
```diff
@@ -0,0 +1,242 @@
+for which you have a known fixed input and which do not run continuously.
+
+We have these different execution modes because `BATCH` execution allows some
+additional optimizations that we can only do when we now that our input is
```
```suggestion
additional optimizations that we can only do when we know that our input is
```

alpinegizmo commented on docs/dev/datastream_execution_mode.md:
```diff
@@ -0,0 +1,242 @@
+The BATCH execution mode can only be used for Jobs/Flink Programs that are
+_bounded_. Boundedness is a property of a data source that tells us whether all
+the input coming from that source is known before execution or whether new data
+will show up, potentially for forever. A job, in turn, is bounded if all its
```
```suggestion
will show up, potentially indefinitely. A job, in turn, is bounded if all its
```

alpinegizmo commented on docs/dev/datastream_execution_mode.md:
```diff
@@ -0,0 +1,242 @@
+_bounded_. Boundedness is a property of a data source that tells us whether all
+the input coming from that source is known before execution or whether new data
+will show up, potentially for forever. A job, in turn, is bounded if all its
+sources are bounded and unbounded otherwise.
```
```suggestion
sources are bounded, and unbounded otherwise.
```

alpinegizmo commented on docs/dev/datastream_execution_mode.md:
```diff
@@ -0,0 +1,242 @@
+
+Another term that is typically used for bounded sources is _batch source_. Or,
+we can say that we are working with a batch data set or a batch of data. Some
+typical other terms for an unbounded source are _continuous source_, _streaming
```
```suggestion other terms for an unbounded source are _continuous source_, _streaming
```

dawidwys replied:
Shall we just drop the paragraph whatsoever? Let's maybe establish a single term for it?

aljoscha replied:
Ok, I'm dropping it. Also fine for you, @alpinegizmo ?

alpinegizmo replied:
👍 for dropping it

sjwiesman replied:
Drop it

alpinegizmo commented on docs/dev/datastream_execution_mode.md:
```diff
@@ -0,0 +1,242 @@
+
+One obvious outlier case is when you want to use a bounded job to bootstrap
+some job state that you then want to use in an unbounded job. For example, by
+running a bounded job using STREAMING mode, taking a savepoint and then
```
```suggestion
running a bounded job using STREAMING mode, taking a savepoint, and then
```

alpinegizmo commented on docs/dev/datastream_execution_mode.md:
```diff
@@ -0,0 +1,242 @@
+ - `BATCH`: Batch-style execution on the DataStream API
+ - `AUTOMATIC`: Let the system decide based on the boundedness of the sources
+
+ This can be configured either in the `flink-conf.yaml`, via command line
```
```suggestion
 This can be configured either in `flink-conf.yaml`, via command line
```

alpinegizmo commented on docs/dev/datastream_execution_mode.md:
```diff
@@ -0,0 +1,242 @@
+that are running in parallel on multiple TaskManagers but we will only use the
+term _task_ here.
+
+Task scheduling and the network shuffle work different for BATCH execution mode
```
```suggestion
Task scheduling and the network shuffle work differently for BATCH execution mode
```

alpinegizmo commented on docs/dev/datastream_execution_mode.md:
```diff
@@ -0,0 +1,242 @@
+
+### Failure Recovery
+
+In STREAMING execution mode, Flink uses checkpoints for failure recover. Take a
```
```suggestion
In STREAMING execution mode, Flink uses checkpoints for failure recovery. Take a
```

alpinegizmo commented on docs/dev/datastream_execution_mode.md:
```diff
@@ -0,0 +1,242 @@
+
+DataStreamSource<String> source = env.fromElements(...);
+
+source.name("source")
+	.map(...).name("map1")
+	.map(...).name("map2")
+	.rebalance()
+	.map(...).name("map3")
+	.map(...).name("map4")
+	.keyBy((value) -> value)
+	.map(...).name("map5")
+	.map(...).name("map6")
+	.sinkTo(...).name("sink");
```
A diagram of an execution graph for this topology would be helpful.

aljoscha replied:
I was afraid someone would suggest this, but yes, I assumed that I would have to do this... 😅🙈

aljoscha replied:
I added a figure.

alpinegizmo commented on docs/dev/datastream_execution_mode.md:
```diff
@@ -0,0 +1,242 @@
+execution mode when your program is unbounded because only this mode is general
+enough to be able to deal with continuous data streams.
+
+TODO: Should we even go into this?
```
I'm in favor of keeping at least some mention of the bootstrapping case. We could also mention that it makes sense to write tests that run in STREAMING mode against bounded sources.

dawidwys replied:
+1 for outlining the testing case

aljoscha replied:
Will do

dawidwys commented on docs/dev/datastream_execution_mode.md:
```diff
@@ -0,0 +1,242 @@
+We have these different execution modes because `BATCH` execution allows some
+additional optimizations that we can only do when we now that our input is
+bounded. For example, different join/aggregation strategies can be used, in
+additional to a different shuffle implementation that allows more efficient
```
```suggestion
addition to a different shuffle implementation that allows more efficient
```

dawidwys commented on docs/dev/datastream_execution_mode.md:
```diff
@@ -0,0 +1,242 @@
+Flink jobs consist of different operations that are connected together in a
+dataflow graph. The system decides how to schedule the execution of these
+operations on different processes/machines (TaskManager) and how data is
+shuffled (sent) betwixt them.
```
I appreciate learning a new word `betwixt`, but I think it is nicer towards our readers, which are very often not native speakers, to stick to the simpler `between`.

aljoscha replied:
I just put it there for someone to find it. 😊

alpinegizmo replied:
I was bemused to encounter such a fine word, and correctly assumed I didn't need to be the one to react to it. 😏

aljoscha replied:
🧐

dawidwys commented on docs/dev/datastream_execution_mode.md:
```diff
@@ -0,0 +1,242 @@
+
+Network shuffles are _pipelined_, meaning that records are immediately sent to
+downstream tasks, with some buffering on the network layer. Again, this enables
+low-latency processing.
```
It's good the way it is now. Just me nitpicking.

I'd say it is not (only) about low-latency. It's (also) because of the unbounded characteristic of the processing. There is no point in the processing that we can say e.g. an operator will not see any more records and we can drop its state. We must keep its state on all the time because it might change at any point in time in the future.

sjwiesman replied:
I agree, I think continuous is more important to highlight here than latency

aljoscha replied:
I expanded this paragraph. PTAL.

dawidwys commented on docs/dev/datastream_execution_mode.md:
```diff
@@ -0,0 +1,242 @@
+downstream tasks to read them after upstream tasks have already gone off line.
+This will increase the latency of processing but comes with other interesting
+properties. For one, this allows Flink to backtrack to the latest available
+results when a failure happens instead of restarting the whole job.
```
Shall we also mention we can run the job with less resources than the sum of all operators required resources?

aljoscha replied:
Will do

aljoscha commented:
Thanks for the speedy review! And I really appreciate the suggestions that I can just apply right here in Github.

I pushed a commit that should address most comments.

sjwiesman commented on docs/dev/datastream_execution_mode.md:
```diff
@@ -0,0 +1,246 @@
+can choose depending on the requirements of your use case and the
+characteristics of your job.
+
+There is the "classic" execution behavior of the DataStream API which we call
```
```suggestion
There is the "classic" execution behavior of the DataStream API, which we call
```

sjwiesman commented on docs/dev/datastream_execution_mode.md:
```diff
@@ -0,0 +1,246 @@
+processing frameworks such as MapReduce. This should be used for bounded jobs
+for which you have a known fixed input and which do not run continuously.
+
+We have these different execution modes because `BATCH` execution allows some
+additional optimizations that we can only do when we know that our input is
+bounded. For example, different join/aggregation strategies can be used, in
+addition to a different shuffle implementation that allows more efficient
+failure recovery behavior. We will go into some of the details of the execution
+behavior below.
```
```suggestion
Apache Flink's unified approach to stream and batch processing means that a DataStream application executed
over bounded input will produce the same results regardless of the configured execution mode.
By enabling `BATCH` execution, we allow Flink to apply additional optimizations that we can only do work when we know that our input is bounded. For example, different join/aggregation strategies can be used, in
addition to a different shuffle implementation that allows more efficient
failure recovery behavior. We will go into some of the details of the execution
behavior below.
```

sjwiesman commented on docs/dev/datastream_execution_mode.md:
```diff
@@ -0,0 +1,246 @@
+
+## When can/should I use BATCH execution mode?
+
+The BATCH execution mode can only be used for Jobs/Flink Programs that are
```
```suggestion
The `BATCH` execution mode can only be used for Jobs/Flink Programs that are
```

sjwiesman commented on docs/dev/datastream_execution_mode.md:
```diff
@@ -0,0 +1,246 @@
+will show up, potentially indefinitely. A job, in turn, is bounded if all its
+sources are bounded, and unbounded otherwise.
+
+STREAMING execution mode, on the other hand, can be used for both bounded and
```
```suggestion
`STREAMING` execution mode, on the other hand, can be used for both bounded and
```

sjwiesman commented on docs/dev/datastream_execution_mode.md:
```diff
@@ -0,0 +1,246 @@
+typical other terms for an unbounded source are _continuous source_, _streaming
+source_, _stream_, or _infinite stream_.
+
+As a rule of thumb, you should be using BATCH execution mode when your program
```
```suggestion
As a rule of thumb, you should be using `BATCH` execution mode when your program
```

sjwiesman commented on docs/dev/datastream_execution_mode.md:
```diff
@@ -0,0 +1,246 @@
+source_, _stream_, or _infinite stream_.
+
+As a rule of thumb, you should be using BATCH execution mode when your program
+is bounded because this will be more efficient. You have to use STREAMING
```
```suggestion
is bounded because this will be more efficient. You have to use `STREAMING`
```

sjwiesman commented on docs/dev/datastream_execution_mode.md:
```diff
@@ -0,0 +1,246 @@
+
+One obvious outlier case is when you want to use a bounded job to bootstrap
+some job state that you then want to use in an unbounded job. For example, by
+running a bounded job using STREAMING mode, taking a savepoint, and then
```
```suggestion
running a bounded job using `STREAMING` mode, taking a savepoint, and then
```

sjwiesman commented on docs/dev/datastream_execution_mode.md:
```diff
@@ -0,0 +1,246 @@
+running a bounded job using STREAMING mode, taking a savepoint, and then
+restoring that savepoint on an unbounded job. This is a very specific use case
+and one that might soon become obsolete when we allow producing a savepoint as
+additional output of a BATCH execution job.
```
```suggestion
additional output of a `BATCH` execution job.
```

sjwiesman commented on docs/dev/datastream_execution_mode.md:
```diff
@@ -0,0 +1,246 @@
+and one that might soon become obsolete when we allow producing a savepoint as
+additional output of a BATCH execution job.
+
+Another case where you might run a bounded job using STREAMING mode is when
```
```suggestion
Another case where you might run a bounded job using `STREAMING` mode is when
```

sjwiesman commented on docs/dev/datastream_execution_mode.md:
```diff
@@ -0,0 +1,246 @@
+refer to the FLIPs that introduced this feature:
+[FLIP-134](https://cwiki.apache.org/confluence/x/4i94CQ) and
+[FLIP-140](https://cwiki.apache.org/confluence/x/kDh4CQ). As well as the
+documentation about [task scheduling (TODO)](<TODO>).
```
reminder for too

sjwiesman commented on docs/dev/datastream_execution_mode.md:
```diff
@@ -0,0 +1,246 @@
+
+#### STREAMING Execution Mode
+
+In STREAMING execution mode, all tasks need to be online/running all the time.
```
```suggestion
In `STREAMING` execution mode, all tasks need to be online/running all the time.
```

sjwiesman commented on docs/dev/datastream_execution_mode.md:
```diff
@@ -0,0 +1,246 @@
+
+In STREAMING execution mode, all tasks need to be online/running all the time.
+This allows Flink to immediately process new records through the whole
+pipeline, which we need for low-latency stream processing. This also means that
```
```suggestion
pipeline, which we need for continuous and low-latency stream processing. This also means that
```
?

sjwiesman commented on docs/dev/datastream_execution_mode.md:
```diff
@@ -0,0 +1,246 @@
+---
+title: "Execution Mode (Batch/Streaming)"
```
Can you add "Execution Mode" to the glossary? 

Also, does it make sense to add a note to the `dev/batch/index.md` directing new applications to use bounded datastream or table? Or are we not there yet?

aljoscha replied:
I pushed a commit that adds it and I'm also recommending to use the DataStream API in the DataSet docs. I think it's time for a bold move... 😅

sjwiesman commented on docs/dev/datastream_execution_mode.md:
```diff
@@ -0,0 +1,246 @@
+The execution mode can be configured via the `execution.runtime-mode` setting.
+There are three possible values:
+
+ - `STREAMING`: The classic DataStream execution mode
```
```suggestion
 - `STREAMING`: The classic DataStream execution mode (default)
```

I'm assuming `STREAMING` is the default. If not, please mark the correct one.

sjwiesman commented on docs/dev/datastream_execution_mode.md:
```diff
@@ -0,0 +1,246 @@
+
+### Failure Recovery
+
+In STREAMING execution mode, Flink uses checkpoints for failure recovery. Take a
```
```suggestion
In `STREAMING` execution mode, Flink uses checkpoints for failure recovery. Take a
```

sjwiesman commented on docs/dev/datastream_execution_mode.md:
```diff
@@ -0,0 +1,246 @@
+
+One of the characteristics of checkpointing for failure recovery is that Flink
+will restart all the running tasks from a checkpoint in case of a failure. This
+can be more costly than what we have to do in BATCH mode (as explained below),
```
```suggestion
can be more costly than what we have to do in `BATCH` mode (as explained below),
```

sjwiesman commented on docs/dev/datastream_execution_mode.md:
```diff
@@ -0,0 +1,246 @@
+One of the characteristics of checkpointing for failure recovery is that Flink
+will restart all the running tasks from a checkpoint in case of a failure. This
+can be more costly than what we have to do in BATCH mode (as explained below),
+which is one of the reasons that you should use BATCH execution mode if your
```
```suggestion
which is one of the reasons that you should use `BATCH` execution mode if your
```

sjwiesman commented on docs/dev/datastream_execution_mode.md:
```diff
@@ -0,0 +1,246 @@
+which is one of the reasons that you should use BATCH execution mode if your
+job allows it.
+
+In BATCH execution mode, Flink will try and backtrack to previous processing
```
```suggestion
In `BATCH` execution mode, Flink will try and backtrack to previous processing
```

sjwiesman commented on docs/dev/datastream_execution_mode.md:
```diff
@@ -0,0 +1,246 @@
+
+## Important Considerations
+
+What doesn't work:
+ - broadcast state/pattern
+ - iterations
+ - operations that rely on checkpointing
+ - this includes most "regular" sinks
```
```suggestion
Full unification of batch and stream processing under the DataStream API is still under active development. 
That means there are still a number of DataStream features that only work under the `STREAMING` execution mode. 

 - broadcast state/pattern
 - iterations
 - operations that rely on checkpointing
 - this includes most "regular" sinks
```

Aljoscha, I'll need you to expand on the bullet points.

aljoscha replied:
Sorry, I forgot to mention that this documentation is WIP and that Dawid has some incoming changes for this part. 🙈

aljoscha replied:
Sorry, Teach

tillrohrmann commented on docs/dev/datastream_execution_mode.md:
```diff
@@ -0,0 +1,241 @@
+This will increase the latency of processing but comes with other interesting
+properties. For one, this allows Flink to backtrack to the latest available
+results when a failure happens instead of restarting the whole job. Another
+side effect is that BATCH jobs can execute on fewer resources (in terms of
```
```suggestion
side effect is that `BATCH` jobs can execute on fewer resources (in terms of
```

tillrohrmann commented on docs/dev/datastream_execution_mode.md:
```diff
@@ -0,0 +1,241 @@
+sequentially one after the other.
+
+TaskManagers will keep intermediate results at least as long as downstream
+tasks have not consumed them. After that, they will be kept for as long as
+space allows in order to allow the aforementioned backtracking to earlier
```
I think technically we keep them until all consuming pipelined regions have produced their outputs.

aljoscha replied:
I added an explanation in parentheses but right now it will point to the non-existing documentation about pipelined regions. That's another TODO. 😅

tillrohrmann reviewed (commented):
Thanks for creating this documentation PR @aljoscha. I had a couple of minor comments.

aljoscha commented on docs/dev/datastream_execution_mode.md:
```diff
@@ -0,0 +1,241 @@
+ - `BATCH`: Batch-style execution on the DataStream API
+ - `AUTOMATIC`: Let the system decide based on the boundedness of the sources
+
+ This can be configured either in `flink-conf.yaml`, via command line
```
You're right, I'll take out the mention of `flink-conf.yaml`. 👌

aljoscha commented on docs/dev/datastream_execution_mode.md:
```diff
@@ -0,0 +1,241 @@
+This section provides an overview of the execution behavior of BATCH execution
+mode and contrasts it with STREAMING execution mode. For more details, please
+refer to the FLIPs that introduced this feature:
+[FLIP-134](https://cwiki.apache.org/confluence/x/4i94CQ) and
```
I click the "Share" button/menu at the top and take the link from there. I'm hoping that this is stable, because just taking a URL like https://cwiki.apache.org/confluence/display/FLINK/FLIP-134%3A+Batch+execution+for+the+DataStream+API is not stable when someone renames the title. There's also the option of using https://s.apache.org/FLIP-134, which is a shortlink I manually created.

aljoscha commented:
I addressed more comments and also added the `important considerations` section by Dawid. Could you please take another look?

aljoscha commented:
I now also added sections by Klou and a section about state backends. Now all the content is theoretically in.

kl0u commented on docs/dev/datastream_execution_mode.md:
```diff
@@ -0,0 +1,377 @@
+
+Apache Flink's unified approach to stream and batch processing means that a
+DataStream application executed over bounded input will produce the same
+results regardless of the configured execution mode.  By enabling `BATCH`
```
remove space in the "  By enabling" (it has two spaces).

aljoscha replied:
👍

kl0u commented on docs/dev/datastream_execution_mode.md:
```diff
@@ -0,0 +1,377 @@
+DataStream application executed over bounded input will produce the same
+results regardless of the configured execution mode.  By enabling `BATCH`
+execution, we allow Flink to apply additional optimizations that we can only do
+work when we know that our input is bounded. For example, different
```
remove the "work": "to apply additional optimizations that we can only do
work when we know that our input is bounded" -> "to apply additional optimizations that we can only do
when we know that our input is bounded"

aljoscha replied:
🆗

kl0u commented on docs/dev/datastream_execution_mode.md:
```diff
@@ -0,0 +1,377 @@
+execution, we allow Flink to apply additional optimizations that we can only do
+work when we know that our input is bounded. For example, different
+join/aggregation strategies can be used, in addition to a different shuffle
+implementation that allows more efficient failure recovery behavior. We will go
```
"that allows more efficient failure recovery behavior" -> "that allows more efficient scheduling and failure recovery."

aljoscha replied:
👌

kl0u commented on docs/dev/datastream_execution_mode.md:
```diff
@@ -0,0 +1,377 @@
+execution mode when your program is unbounded because only this mode is general
+enough to be able to deal with continuous data streams.
+
+One obvious outlier case is when you want to use a bounded job to bootstrap
```
"One obvious outlier case is when you want" -> "One obvious outlier is when you want"

aljoscha replied:
fixing

kl0u commented on docs/dev/datastream_execution_mode.md:
````diff
@@ -0,0 +1,377 @@
+config.set(ExecutionOptions.RUNTIME_MODE, RuntimeExecutionMode.BATCH);
+StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment(config);
+ ```
+
````
I would add: 

```
or 

```java
StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment(config);
env.setRuntimeMode(RuntimeExecutionMode.BATCH);
 ```

<span class="label label-danger">Important Note</span>: We recommend users to NOT use this method
but set the `execution.runtime-mode` using the command-line when submitting the application. Keeping 
the application code configuration-free allows for more flexibility as the same application will be
able to be executed in any execution mode.

```

aljoscha replied:
Thanks, that's a good suggestion!

kl0u commented on docs/dev/datastream_execution_mode.md:
```diff
@@ -0,0 +1,377 @@
+
+Flink jobs consist of different operations that are connected together in a
+dataflow graph. The system decides how to schedule the execution of these
+operations on different processes/machines (TaskManager) and how data is
```
TaskManager -> TaskManagers

aljoscha replied:
fixing

kl0u commented on docs/dev/datastream_execution_mode.md:
```diff
@@ -0,0 +1,377 @@
+that are running in parallel on multiple TaskManagers but we will only use the
+term _task_ here.
+
+Task scheduling and the network shuffle work differently for `BATCH` execution
```
"Task scheduling and the network shuffle" -> "Task scheduling and network shuffles"

"differently for `BATCH` execution mode and `STREAMING` execution mode" -> "differently for `BATCH` and `STREAMING` execution mode"

aljoscha replied:
fixing

kl0u commented on docs/dev/datastream_execution_mode.md:
```diff
@@ -0,0 +1,377 @@
+
+TaskManagers will keep intermediate results at least as long as downstream
+tasks have not consumed them. (Technically, they will be kept until the
+consuming [pipelined regions TODO](<TODO>) have produced their output in turn.)
```
I would remove the "in turn".

aljoscha replied:
fixing

kl0u commented on docs/dev/datastream_execution_mode.md:
```diff
@@ -0,0 +1,377 @@
+
+### Checkpointing
+
+As explained [above](#failure-recovery), fault tolerance for batch programs
```
"fault tolerance for batch programs does not use checkpointing" -> "failure recovery for batch programs
does not use checkpointing". 

I would suggest that because technically it is not fault tolerance as much as leveraging caching to speed up the catch up phase of reprocessing.

aljoscha replied:
good catch!

kl0u reviewed (commented):
Thanks for the work @aljoscha , I left some comments in the PR. Feel free to integrate whichever you agree with.

sjwiesman reviewed (commented):
+1 to @kl0u 's comments and few others

tillrohrmann commented on docs/dev/datastream_execution_mode.md:
```diff
@@ -0,0 +1,377 @@
+`STREAMING` execution mode, on the other hand, can be used for both bounded and
+unbounded jobs.
+
+As a rule of thumb, you should be using `BATCH` execution mode when your program
+is bounded because this will be more efficient. You have to use `STREAMING`
+execution mode when your program is unbounded because only this mode is general
+enough to be able to deal with continuous data streams.
```
Here it sounds as if it does not really matter whether to choose `BATCH` or `STREAMING` for a bounded job from a correctness perspective. However, the `FileSink` won't commit the in-progress files at the end of the program when using the `STREAMING` execution mode. It might be worthwhile to document this behaviour somewhere.

aljoscha replied:
Yes, this is unfortunate. Though the fact that we cannot do checkpoints as soon as at least one task has finished, which in turn means that we can't get a "final" checkpoint has been a feature/bug of DataStream execution since the beginning. I wouldn't document it here but we can think about adding this to a general "caveats" section. I'm sure there would be other corner cases that are worth documenting 😅

tillrohrmann reviewed (commented):
I had one more comment about choosing the `BATCH` vs. the `STREAMING` execution mode.

aljoscha commented on docs/dev/datastream_execution_mode.md:
```diff
@@ -0,0 +1,377 @@
+not known in advance. In these settings, requirements may change over time
+depending on the incoming data.
+
+In the batch world though, we believe that such use-cases do not make much
+sense, as the input (both the elements and the control stream) are static and
+known in advance.
```
I'm afraid so. But it's not that nice. We do want to add proper support for broadcast input in the next release, though.

tillrohrmann commented on docs/dev/datastream_execution_mode.md:
```diff
@@ -0,0 +1,377 @@
+
+As explained [above](#failure-recovery), fault tolerance for batch programs
+does not use checkpointing. Recovery happens by fully replaying the streams.
+This is possible because inputs are bounded.  This pushes the cost more towards
+the recovery, but makes the regular processing cheaper, because it avoids
+checkpoints.
```
The last paragraph of `failure recovery` reads as if `BATCH` execution improves the overall execution time of jobs but here it reads a bit differently. Concretely that `BATCH` recoveries are more costly than `STREAMING` recoveries.

aljoscha replied:
Good catch! @dawidwys what was the original intention here?

dawidwys replied:
My intention was to briefly remind the batch failure recovery model. For that I actually reused the description from: https://ci.apache.org/projects/flink/flink-docs-master/concepts/stateful-stream-processing.html#state-and-fault-tolerance-in-batch-programs

dawidwys replied:
With the description in the `failure recovery` section, we can probably drop the first paragraph and start with the second one:

>It is important to remember that because there are no checkpoints, as described [above](#failure-recovery), certain ...

BTW shall we update it in `#stateful-stream-processing.html#state-and-fault-tolerance-in-batch-programs`? It is not easy to tell which model recovers "faster" as it very much depends on the state size, number of records to replay, number of tasks to recover etc.

aljoscha replied:
Oh boy, this text was first added in 2016: https://github.com/apache/flink/blame/b04d51a129a3341887e7a0866557c9871f58e94c/docs/concepts/concepts.md. I copied it to the current concepts section from there. That's not at all up-to-date anymore.

That section needs an overhaul or should be removed because it's also misleading for `DataSet` programs or Table/SQL batch programs.

aljoscha replied:
For this documentation here I think we can go with @dawidwys's suggestion and just drop that paragraph because I added some text about that above.

dawidwys reviewed (approved):
+1 from my side for these changes

aljoscha commented:
I believe I addressed all comments. Please take another look. If there's no objection I would merge this by tomorrow because this PR/discussion is growing a bit unwieldy. Anything else we want to add we can still add later.

aljoscha commented:
Thanks for the reviews! I merged this now.
